File: rlwe-challenges/parsers/java/ChallInstParser.java

// This file prints the contents of challenge and instance files.
// The path to a .challenge file must be provided; the path to a corresponding
// instance file may also be provided. The challenge contents are printed, and
// the instance contents are printed if a path was provided.
// See comments in the code for examples of how to access message contents.

// The included jar file was obtained from http://search.maven.org/remotecontent?filepath=com/google/protobuf/protobuf-java/2.6.1/protobuf-java-2.6.1.jar
// > javac -cp protobuf-java-2.6.1.jar crypto/proto/lol/*.java
// > javac -cp .:protobuf-java-2.6.1.jar crypto/proto/RLWE/*.java
// > javac -cp .:protobuf-java-2.6.1.jar crypto/proto/RLWE/challenges/*.java
// > javac -cp .:protobuf-java-2.6.1.jar ChallInstParser.java
// > java  -cp .:protobuf-java-2.6.1.jar ChallInstParser path/to/.challenge [path/to/.instance]

import crypto.proto.RLWE.challenges.*;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.PrintStream;

class ChallInstParser {
	public static void main(String[] args) throws Exception {

		// read a challenge file
		Challenges.Challenge chall =
			Challenges.Challenge.parseFrom(new FileInputStream(args[0]));
		System.out.println(chall); // print all contents
		// some examples showing how to access message contents
		/*
		System.out.println(chall.getNumInstances()); // access a field
		System.out.println(chall.getParamsCase()); // an enum showing which type of parameter the message contains
		switch (chall.getParamsCase()) {
			case CPARAMS: {
        System.out.println(chall.getCparams().getM());
        break;
      }
      case DPARAMS: {
        System.out.println(chall.getDparams().getNumSamples());
        break;
      }
      case RPARAMS: {
        System.out.println(chall.getRparams().getQ());
        break;
      }
		}
		*/
		if (args.length == 2) {
			// also read an instance file
			if (chall.hasCparams()) {
				Challenges.InstanceCont inst =
					Challenges.InstanceCont.parseFrom(new FileInputStream(args[1]));
				System.out.println(inst); // print all contents
				// it is easy to access any member of the parsed message
        //System.out.println(inst.getSamples(0).getA().getXs(0));
			}
			else if (chall.hasDparams()) {
				Challenges.InstanceDisc inst =
					Challenges.InstanceDisc.parseFrom(new FileInputStream(args[1]));
				System.out.println(inst); // print all contents
			}
			else if (chall.hasRparams()) {
				Challenges.InstanceRLWR inst =
					Challenges.InstanceRLWR.parseFrom(new FileInputStream(args[1]));
				System.out.println(inst); // print all contents
			}
			else {
				System.out.println("Challenge parse error.");
				System.exit(0);
			}
		}
	}
}

File: rlwe-challenges/parsers/python/ChallInstParser.py
#!/usr/bin/python

# Parses RLWE challenge/instance files:
# takes the path to a .challenge file first, and an optional second path to a
# corresponding .instance file. Prints the contents of the .challenge file
# and the .instance file, if a path is provided.
# See comments in the code for examples of how to access message contents.
#
# You will need the "protobuf" python package to run this code, which you can
# install with "easy_install protobuf"

# You can find more detailed information about parsing messages in python here:
# https://developers.google.com/protocol-buffers/docs/reference/python-generated
import sys
import Challenges_pb2

def parse_challenge(challenge_path):
  if challenge_path is None:
    return
  challenge = None
  with open(challenge_path) as f:
    challenge = Challenges_pb2.Challenge()
    challenge.ParseFromString(f.read())
  return challenge

def parse_instance(challenge, instance_path):
  if challenge is None or instance_path is None:
    return
  instance = None

  with open(instance_path) as f:
    if challenge.cparams is not None:
      instance = Challenges_pb2.InstanceCont()
    elif challenge.dparams is not None:
      instance = Challenges_pb2.InstanceDisc()
    elif challenge.rparams is not None:
      instance = Challenges_pb2.InstanceRLWR()
    else:
      return
    instance.ParseFromString(f.read())
  return instance

if __name__ == "__main__":

  if len(sys.argv) < 2 or len(sys.argv) > 3:
    print "Usage:", sys.argv[0], "path/to/.challenge [path/to/.instance]"
    sys.exit(-1)

  challenge = parse_challenge(sys.argv[1])

  if challenge is None:
    print "Could not parse the challenge parameters."
    sys.exit(-1)
  print challenge # print all contents
  # examples of how to access message contents
'''
  print challenge.numInstances # access a field
  print challenge.WhichOneof("params") # shows which type of parameter the message contains
  print {'cparams': challenge.cparams.m,
         'dparams': challenge.dparams.numSamples,
         'rparams': challenge.rparams.q}[challenge.WhichOneof("params")]
'''
  if len(sys.argv) == 3:
    instance = parse_instance(challenge, sys.argv[2])
    if instance is None:
      print "Could not parse instance."
      sys.exit(-1)
    print instance # print all contents
    # it is easy to access any member of the parsed message
    #print instance.samples[0].a.xs[0]

  sys.exit(0)